lib/LaTeXML/Util/Pathname.pm

```diff
@@ -358,26 +357,41 @@ sub candidate_pathnames {
         push(@exts, '.' . $ext); } } }
   push(@exts, '') unless @exts;
 
-  my @paths = ();
+  my @paths        = ();
+  my @nocase_paths = ();
+  # Precompute the intended regexes to apply
+  my @regexes = ();
+  foreach my $ext (@exts) {
+    if ($name eq '*') {    # Unfortunately, we've got to test the file system NOW...
+      if ($ext eq '.*') {    # everything, except hidden files
+        push(@regexes, [qr/^[^.]/, qr/^[^.]/]); }
+      else {
+        push(@regexes, [qr/\Q$ext\E$/i,
+            qr/\Q$ext\E$/]); } }
+    elsif ($ext eq '.*') {
+      push(@regexes, [qr/^\Q$name\E\.\w+$/i,
+          qr/^\Q$name\E\.\w+$/]); }
+    else {
+      push(@regexes, [qr/^\Q$name\E\Q$ext\E$/i,
+          qr/^\Q$name\E\Q$ext\E$/]); } }
   # Now, combine; precedence to leading directories.
   foreach my $dir (@dirs) {
-    foreach my $ext (@exts) {
-      if ($name eq '*') {    # Unfortunately, we've got to test the file system NOW...
-        if ($ext eq '.*') {    # everything
-          opendir(DIR, $dir) or next;
-          push(@paths, map { pathname_concat($dir, $_) } grep { !/^\./ } readdir(DIR));
-          closedir(DIR); }
-        else {
-          opendir(DIR, $dir) or next;    # ???
-          push(@paths, map { pathname_concat($dir, $_) } grep { /\Q$ext\E$/ } readdir(DIR));
-          closedir(DIR); } }
-      elsif ($ext eq '.*') {             # Unfortunately, we've got to test the file system NOW...
-        opendir(DIR, $dir) or next;      # ???
-        push(@paths, map { pathname_concat($dir, $_) } grep { /^\Q$name\E\.\w+$/ } readdir(DIR));
-        closedir(DIR); }
-      else {
-        push(@paths, pathname_concat($dir, $name . $ext)); } } }
-  return @paths; }
+    opendir(DIR, $dir) or next;
+    my @dir_files = readdir(DIR);
+    closedir(DIR);
+    for my $local_file (@dir_files) {
+      for my $regex_pair (@regexes) {
+        my ($i_regex, $regex) = @$regex_pair;
+        if ($local_file =~ m/$i_regex/) {
+          my $full_file = pathname_concat($dir, $local_file);
+          push(@nocase_paths, $full_file);
+          if ($local_file =~ m/$regex/) {
+            # if we are only interested in the first match, return it:
+            return ($full_file) if $options{findfirst};
+            push(@paths, $full_file); } } } } }
+  # Fallback: if no strict matches were found, return any existing case-insensitive matches
+  # Defer the -f check until we are sure we need it, to keep the usual cases fast.
+  return @paths ? @paths : grep { -f $_ } @nocase_paths; }
```

Comment:
But again, we don't really need the `-f` test, since the files do exist in the directory; it's just that they don't have the same case as the request.

Reply:
... it's now officially my blindspot. patch incoming

Reply:
...patch landed